src/Parsers/ASTSelectQuery.cpp

```diff
@@ -125,6 +128,18 @@ void ASTSelectQuery::formatImpl(const FormatSettings & s, FormatState & state, F
     if (group_by_with_cube)
         s.ostr << (s.hilite ? hilite_keyword : "") << s.nl_or_ws << indent_str << (s.one_line ? "" : "    ") << "WITH CUBE" << (s.hilite ? hilite_none : "");
 
+    if (group_by_with_grouping_sets)
+    {
+        frame.surround_each_list_element_with_parens = true;
+        s.ostr << (s.hilite ? hilite_keyword : "") << s.nl_or_ws << indent_str << (s.one_line ? "" : "    ") << "GROUPING SETS" << (s.hilite ? hilite_none : "");
+        s.ostr << " (";
+        s.one_line
+        ? groupBy()->formatImpl(s, state, frame)
+        : groupBy()->as<ASTExpressionList &>().formatImplMultiline(s, state, frame);
+        s.ostr << ")";
+        frame.surround_each_list_element_with_parens = false;
+    }
```

Comment:
Formatting looks odd:
```sql
SELECT
    key1,
    key2,
    sum(value)
FROM data
GROUP BY
    GROUPING SETS (
    ( key1),
    ( key2))
```

So looks like:
- `surround_each_list_element_with_parens` should not be set
- extra parenthesis is not required

So something like this:

```c
if (group_by_with_grouping_sets)
{
    s.ostr << (s.hilite ? hilite_keyword : "") << s.nl_or_ws << indent_str << (s.one_line ? "" : "    ") << "GROUPING SETS" << (s.hilite ? hilite_none : "");
    s.one_line
    ? groupBy()->formatImpl(s, state, frame)
    : groupBy()->as<ASTExpressionList &>().formatImplMultiline(s, state, frame);
}
```

Reply:
> extra parenthesis is not required

Why? This is part of GROUPING SETS syntax.

But I agree I should remove new line before GROUPING SETS and redundant whitespace before the first element in grouping set.

Reply:
>Why? This is part of GROUPING SETS syntax.

I don't see why they are required, both queries works fine:
- `SELECT key1, key2, sum(value) FROM data GROUP BY GROUPING SETS (( key1), ( key2))`
- `SELECT key1, key2, sum(value) FROM data GROUP BY GROUPING SETS key1, key2`

Plus `GROUP BY` keys are not surrounded with extra parenthesis.

Reply:
Because grouping sets support several grouping commands. 
This is a valid syntax:
```
GROUPING SETS ((key1), (key2), (key1, key2), ())
```
It should work similar to Postgres.

Reply:
Got it, thanks for the clarifications.